stan/math/rev/fun/eigenvectors_sym.hpp

```diff
@@ -14,86 +14,42 @@
-        vari_ref_A_(ChainableStack::instance_->memalloc_.alloc_array<vari *>(
-            A.rows() * A.cols())),
-        vari_ref_v_(ChainableStack::instance_->memalloc_.alloc_array<vari *>(
-            A.rows() * A.cols())) {
-    using Eigen::Map;
-
-    Map<matrix_d> Ad(A_, M_, M_);
-    Map<matrix_d> wd(w_, M_, 1);
-    Map<matrix_d> vd(v_, M_, M_);
-    Ad = A.val();
-    Eigen::SelfAdjointEigenSolver<matrix_d> solver(Ad);
-    wd = solver.eigenvalues();
-    vd = solver.eigenvectors();
-
-    Map<matrix_vi>(vari_ref_A_, M_, M_) = A.vi();
-    Map<matrix_vi>(vari_ref_v_, M_, M_)
-        = vd.unaryExpr([](double x) { return new vari(x, false); });
-  }
-
-  /**
-   * Reverse mode differentiation algorithm reference:
-   *
-   * Mike Giles. An extended collection of matrix derivative results for
-   * forward and reverse mode AD.  Jan. 2008.
-   *
-   * Section 3.1 Eigenvalues and eigenvectors.
-   */
-  virtual void chain() {
-    using Eigen::Map;
-
-    matrix_d adj_v = Map<matrix_vi>(vari_ref_v_, M_, M_).adj();
-    Map<matrix_d> w(w_, M_, M_);
-    Map<matrix_d> v(v_, M_, M_);
-
-    matrix_d f(M_, M_);
-    for (int i = 0; i < M_; i++)
-      for (int j = 0; j < M_; j++)
-        f.coeffRef(j, i) = (i != j ? 1 / (w.coeff(i) - w.coeff(j)) : 0);
-
-    matrix_d adjA = v * f.cwiseProduct(v.transpose() * adj_v) * v.transpose();
-
-    Map<matrix_vi>(vari_ref_A_, M_, M_).adj() += adjA;
-  }
-};
-}  // namespace internal
-
 /**
  * Return the eigenvectors of the specified symmetric matrix.
  * <p>See <code>eigen_decompose()</code> for more information.
+ * @tparam T type of input matrix.
  * @param m Specified matrix.
  * @return Eigenvectors of matrix.
  */
-inline matrix_v eigenvectors_sym(const matrix_v &m) {
-  matrix_d m_eval(value_of_rec(m));
-  check_nonzero_size("eigenvalues_sym", "m", m_eval);
-  check_symmetric("eigenvalues_sym", "m", m_eval);
-  matrix_v res(m.rows(), m.cols());
-  internal::eigenvectors_vari *baseVari = new internal::eigenvectors_vari(m);
-  res.vi()
-      = Eigen::Map<matrix_vi>(baseVari->vari_ref_v_, res.rows(), res.cols());
-  return res;
+template <typename T, require_rev_matrix_t<T>* = nullptr>
+inline auto eigenvectors_sym(const T& m) {
+  check_nonzero_size("eigenvalues_sym", "m", m);
+  check_symmetric("eigenvalues_sym", "m", m);
+
+  using return_t = return_var_matrix_t<T>;
+  auto arena_m = to_arena(m);
+
+  Eigen::SelfAdjointEigenSolver<Eigen::MatrixXd> solver(arena_m.val());
+  arena_t<return_t> eigenvecs = solver.eigenvectors();
+  auto eigenvals = to_arena(solver.eigenvalues());
+
+  reverse_pass_callback([arena_m, eigenvals, eigenvecs]() mutable {
+    const int p = arena_m.val().cols();
+    auto eigenvecs_a = to_arena(eigenvecs.val());
+    auto eigenvecs_a_adj = to_arena(eigenvecs.adj());
+
+    matrix_d f(p, p);
+    for (int i = 0; i < p; ++i)
+      for (int j = 0; j < p; ++j)
+        f.coeffRef(j, i)
+            = (i != j ? 1 / (eigenvals.coeff(i) - eigenvals.coeff(j)) : 0);
```

Comment:
We tend to use i for the rows and j for the columns. And I think you can make this

```cpp
    Eigen::MatrixXd f = (1 / (eigenvals.rowwise().replicate(p).transpose() - eigenvals.rowwise().replicate(p)).array());
    f.diagonal().setZero();
```

We could do more clever things with `selfAdjointView` but idk how to do elementwise multiplication with a `selfAdjointView` which we would need below